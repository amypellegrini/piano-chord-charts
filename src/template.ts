type KeyboardChartFormat = "compact" | "exact";

export type WhiteKey = "C" | "D" | "E" | "F" | "G" | "A" | "B";
export type BlackKey =
  | "C#"
  | "D#"
  | "F#"
  | "G#"
  | "A#"
  | "Db"
  | "Eb"
  | "Gb"
  | "Ab"
  | "Bb";

type RenderOptions = {
  format?: KeyboardChartFormat;
  highlightKeys?: Array<WhiteKey | BlackKey>;
  size?: number;
};

type OffsetKey = "C" | "D" | "F" | "G" | "A";

type OffsetKeyMap = {
  [key in OffsetKey]: number;
};

const keyNames = ["C", "D", "E", "F", "G", "A", "B"];
const whiteKeyWidth = 23;

function renderWhiteKeys(
  height: number,
  highlightKeys?: WhiteKey[],
  size?: number
) {
  const amount = size || 14;

  let keyX = 0;
  let result = "";
  let highlightIdx = 0;

  for (let count = 0; count < amount; count++) {
    const keyName = keyNames[count] || keyNames[count - 7];

    let colour = "#fafafa";

    if (highlightKeys) {
      if (
        highlightIdx < highlightKeys.length &&
        highlightKeys[highlightIdx] === keyName
      ) {
        colour = "#a0c6e8";
        highlightIdx += 1;
      }
    }

    keyX = count * whiteKeyWidth;

    result = result.concat(
      `<rect style="fill:${colour};stroke:black" x="${keyX}" y="0" width="23" height="${height}" ry="3"></rect>\n`
    );
  }

  return result;
}

function renderBlackKeys(
  height: number,
  highlightKeys?: BlackKey[],
  size?: number
) {
  const whiteKeysAmount = size || 14;

  const offsetFromWhiteKeyMap: OffsetKeyMap = {
    C: 14.33333,
    D: 18.66666,
    F: 13.25,
    G: 16.25,
    A: 19.75,
  };

  const flatToSharpMap: { [key: string]: string } = {
    Db: "C#",
    Eb: "D#",
    Gb: "F#",
    Ab: "G#",
    Bb: "A#",
  };

  let result = "";
  let highlightIdx = 0;

  for (let count = 0; count < whiteKeysAmount; count++) {
    const keyName = keyNames[count] || keyNames[count - 7];
    const blackKeyName = keyName + "#";

    let colour = "#222222";
    let highlightKeyName = (highlightKeys && highlightKeys[highlightIdx]) || "";

    if (highlightKeyName.match(/b/)) {
      highlightKeyName = flatToSharpMap[highlightKeyName];
    }

    if (keyName in offsetFromWhiteKeyMap) {
      if (highlightKeys && blackKeyName === highlightKeyName) {
        colour = "#a0c6e8";
        highlightIdx++;
      }

      const offset = offsetFromWhiteKeyMap[keyName as OffsetKey];
      const keyX = count * whiteKeyWidth + offset;

      result = result.concat(
        `<rect style="fill:${colour};stroke:black" x="${keyX}" y="0" width="13" height="${height}" ry="1"></rect>\n`
      );
    }
  }

  return result;
}

function render(options?: RenderOptions) {
  let height = 65;
  let blackKeyHeight = 40;

  const format: KeyboardChartFormat = options?.format || "compact";
  const size: number = options?.size || 14;
  const width: number = size * whiteKeyWidth;

  if (format === "exact") {
    height = 120;
    blackKeyHeight = 80;
  }

  const highlightWhiteKeys = options?.highlightKeys?.filter((key) => {
    return !!key.match(/(A|B|C|D|E|F|G)$/);
  });

  const highlightBlackKeys = options?.highlightKeys?.filter((key) => {
    return !!key.match(/(A#|C#|D#|F#|G#|Db|Eb|Gb|Ab|Bb)$/);
  });

  const defaultContent = `${renderWhiteKeys(
    height,
    highlightWhiteKeys as WhiteKey[],
    size
  )}${renderBlackKeys(
    blackKeyHeight,
    highlightBlackKeys as BlackKey[],
    size
  )}<rect y="-1" width="${width}" height="3"></rect>`;

  const template = `<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
  <svg width="${width + 1}" height="${
    height + 1
  }" viewBox="0 0 ${width} ${height}" version="1.1" xmlns="http://www.w3.org/2000/svg">
    ${defaultContent}
  </svg>`;

  return template;
}

const template = render();

export { render, renderWhiteKeys, renderBlackKeys };
export default template;
