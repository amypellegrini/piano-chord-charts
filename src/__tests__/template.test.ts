import template, { render, renderWhiteKeys, WhiteKey } from "../template";

describe("default chart template", () => {
  it("renders correctly", () => {
    expect(template).toMatchSnapshot();
  });
});

describe("renderWhiteKeys", () => {
  it("renders white keys default SVG content with a given height", () => {
    expect(renderWhiteKeys(120)).toMatchSnapshot();
  });

  it("highlights specific white keys from left to right", () => {
    const highlightKeys: WhiteKey[] = ["C", "E", "G"];
    expect(renderWhiteKeys(120, highlightKeys)).toMatchSnapshot();
  });
});

describe("render", () => {
  it("renders a default compact chart when no options are given", () => {
    const result = render();
    expect(result).toMatchSnapshot();
  });

  it("renders compact and exact versions of the keyboard", () => {
    const defaultRender = render();

    const withCompactOption = render({
      format: "compact",
    });

    const withExactOption = render({
      format: "exact",
    });

    expect(defaultRender).toEqual(withCompactOption);
    expect(defaultRender).not.toEqual(withExactOption);
    expect(withExactOption).toMatchSnapshot();
  });

  it("highlights specific white keys from left to right", () => {
    const result = render({
      highlightKeys: ["C", "E", "G"] as WhiteKey[],
    });

    expect(result).toMatchSnapshot();
  });
});
