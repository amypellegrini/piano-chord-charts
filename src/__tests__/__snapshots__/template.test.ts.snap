// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default chart template renders correctly 1`] = `
"<?xml version=\\"1.0\\" standalone=\\"no\\"?>
<!DOCTYPE svg PUBLIC \\"-//W3C//DTD SVG 1.1//EN\\" \\"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\\">
  <svg width=\\"322\\" height=\\"120\\" viewBox=\\"0 0 322 120\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <rect style=\\"fill:#fafafa;stroke:black\\" x=\\"0\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"23\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"46\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"69\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"92\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"115\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"138\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"161\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"184\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"207\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"230\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"253\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"276\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"299\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"14.33333\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"41.66666\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"82.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"108.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"134.75\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"175.33333\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"202.66666\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"243.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"269.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"295.75\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect y=\\"0\\" width=\\"322\\" height=\\"3\\"></rect>
  </svg>"
`;

exports[`render renders a default chart when no args are given 1`] = `
"<?xml version=\\"1.0\\" standalone=\\"no\\"?>
<!DOCTYPE svg PUBLIC \\"-//W3C//DTD SVG 1.1//EN\\" \\"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\\">
  <svg width=\\"322\\" height=\\"120\\" viewBox=\\"0 0 322 120\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <rect style=\\"fill:#fafafa;stroke:black\\" x=\\"0\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"23\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"46\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"69\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"92\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"115\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"138\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"161\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"184\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"207\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"230\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"253\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"276\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:#fafafa;stroke:black\\" x=\\"299\\" y=\\"0\\" width=\\"23\\" height=\\"120\\" ry=\\"3\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"14.33333\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"41.66666\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"82.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"108.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"134.75\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"175.33333\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"202.66666\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"243.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"269.25\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect style=\\"fill:black;stroke:black\\" x=\\"295.75\\" y=\\"0\\" width=\\"13\\" height=\\"80\\" ry=\\"1\\"></rect>
<rect y=\\"0\\" width=\\"322\\" height=\\"3\\"></rect>
  </svg>"
`;

exports[`render renders gvien content inside the template 1`] = `
"<?xml version=\\"1.0\\" standalone=\\"no\\"?>
<!DOCTYPE svg PUBLIC \\"-//W3C//DTD SVG 1.1//EN\\" \\"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\\">
  <svg width=\\"322\\" height=\\"120\\" viewBox=\\"0 0 322 120\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    text content
  </svg>"
`;
